### PR TITLE
Optimize FormattingHelpers.CountHexDigits using Lzcnt.LeadingZeroCount

### PR DESCRIPTION
**Pull request by EgorBo**
I guess it's now legit to use `Lzcnt` since it's already used by `Decimal` imported from CoreRT recently (see https://github.com/dotnet/coreclr/blob/master/src/System.Private.CoreLib/shared/System/Decimal.DecCalc.cs#L613).

The idea is to calculate leading zeros, for example:
```
ulong number = 0x0000_0000_0000_0FFF;
var lzc = Lzcnt.LeadingZeroCount(number); 
```
gives us 52 bits from overall 64, so (64-52)/4 returns 3 (FFF - 3 hex digits)

The optimization boosts `ToString("X")` by ~5% (and up to 14% for small input values)
```
       Method |  Toolchain |        X |     Mean |
------------- |----------- |--------- |---------:|
    ToStringX |     before |        1 | 18.02 ns |
    ToStringX |      after |        1 | 15.79 ns |
    ToStringX |     before |     3562 | 19.42 ns |
    ToStringX |      after |     3562 | 18.86 ns |
    ToStringX |     before | 14593470 | 21.07 ns |
    ToStringX |      after | 14593470 | 19.68 ns |
```
	
```csharp
public class Benchmarks
{
    [Params(0x1, 0xDEA, 0xDEADBE)]
    public int X { get; set; }

    [Benchmark]
    public string CastAndCount() => X.ToString("X", CultureInfo.InvariantCulture);
}

```

### CODE REVIEW

**Comment by EgorBo** [comment]
@dotnet-bot test OSX10.12 x64 Checked Innerloop Build and Test

File: src/System.Private.CoreLib/shared/System/Buffers/Text/FormattingHelpers.CountDigits.cs [hunk 1]
```diff
@@ -103,8 +104,11 @@ public static int CountDigits(uint value)
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public static int CountHexDigits(ulong value)
         {
-            // TODO: When x86 intrinsic support comes online, experiment with implementing this using lzcnt.
-            // return 16 - (int)((uint)Lzcnt.LeadingZeroCount(value | 1) >> 3);
+            if (Lzcnt.IsSupported && IntPtr.Size == 8)
+            {
+                var right = 64 - (int)Lzcnt.LeadingZeroCount(value | 1);
+                return (right + 3) >> 2;
+            }
```
Comment (stephentoub) [inline src/System.Private.CoreLib/shared/System/Buffers/Text/FormattingHelpers.CountDigits.cs]: Nice. Is this going to cause any problems when it's mirrored over to corefx and used by the Utf8Formatter/Parser logic there? Wondering about assembly references and the like.

Comment (EgorBo) [reply src/System.Private.CoreLib/shared/System/Buffers/Text/FormattingHelpers.CountDigits.cs]: hm.. I am not sure actually probably I need to wrap it with `#if INSIDE_MSCORLIB` or something like that?

Comment (jkotas) [reply src/System.Private.CoreLib/shared/System/Buffers/Text/FormattingHelpers.CountDigits.cs]: It should work fine. You may want to copy the file over the corefx locally and build it to verify that there is no ugly surprise.

File: src/System.Private.CoreLib/shared/System/Buffers/Utilities.cs [hunk 2]
```diff
@@ -14,6 +15,12 @@ internal static int SelectBucketIndex(int bufferSize)
         {
             Debug.Assert(bufferSize >= 0);
 
+            if (Lzcnt.IsSupported && IntPtr.Size == 8)
```
Comment (pentp) [inline src/System.Private.CoreLib/shared/System/Buffers/Utilities.cs]: This works on 32-bit as well, no need to test `IntPtr.Size` here.

Comment (EgorBo) [reply src/System.Private.CoreLib/shared/System/Buffers/Utilities.cs]: @pentp are you sure? I had to add that check to fix few failing CI agents. Like here https://github.com/dotnet/coreclr/blob/master/tests/src/JIT/HardwareIntrinsics/X86/Lzcnt/Lzcnt.cs#L52

Comment (pentp) [reply src/System.Private.CoreLib/shared/System/Buffers/Utilities.cs]: 64-bit test is needed for `Lzcnt(long)` overload, the `int` overload works on both.

Comment (EgorBo) [reply src/System.Private.CoreLib/shared/System/Buffers/Utilities.cs]: ah, makes sense, thanks

**Comment by stephentoub** [comment]
@EgorBo, are you still working on this?

> Do I need to wrap the code here with #if HAS_INTRINSICS

Easiest way to check is probably just to copy your changes here over the corresponding common files in corefx, build, and see if anything breaks.

**Comment by EgorBo** [comment]
@stephentoub well I think it's finished, let me check these changes on top of corefx

**Comment by stephentoub** [comment]
@dotnet-bot test this please

**Comment by stephentoub** [comment]
@dotnet-bot test OSX10.12 x64 Checked Innerloop Build and Test please

**Pull request merged**